### PR TITLE
my vision of SG vision

### PR DESCRIPTION
**Pull request by xDanilcusx**
https://github.com/PvE-CMSS13/PvE-CMSS13/assets/65656972/0c3efe92-ef43-442c-be8b-59b0cd5f2b7e

https://github.com/PvE-CMSS13/PvE-CMSS13/assets/65656972/c7de59cd-220c-4449-a788-aa3aab367979

https://github.com/PvE-CMSS13/PvE-CMSS13/assets/65656972/70374feb-1498-43ea-9d68-36196f855308

*Ignore "only in the dark" part, it's a bit outdated

image

### instead of images that instantly follows every target

### CODE REVIEW

**Comment by fira** [comment]
Can you explicit? Screenshot maybe?

**Comment by xDanilcusx** [comment]
> Can you explicit? Screenshot maybe?

ehh i was doing so in dev channel personally for morrow, i'll post those here from now on

**Comment by xDanilcusx** [comment]
and yeah, those marks sprites (and sound?) are placeholders, gotta need something like one-loop animation of holo-lock onto target

File: code/modules/clothing/glasses/night.dm [hunk 1]
```diff
@@ -86,26 +86,49 @@
...
+
+/obj/effect/target_mark/proc/update_marker_pos(mob/user, atom/target)
+	var/c_view = user.client.view
+	var/view_x_offset = 0
+	var/view_y_offset = 0
+	if(c_view > 7)
+		if(user.client.pixel_x >= 0) view_x_offset = round(user.client.pixel_x/32)
+		else view_x_offset = Ceiling(user.client.pixel_x/32)
+		if(user.client.pixel_y >= 0) view_y_offset = round(user.client.pixel_y/32)
+		else view_y_offset = Ceiling(user.client.pixel_y/32)
```
Comment (harryob) [inline code/modules/clothing/glasses/night.dm:103]:
```suggestion
		if(user.client.pixel_x >= 0) 
			view_x_offset = round(user.client.pixel_x/32)
		else 
			view_x_offset = Ceiling(user.client.pixel_x/32)

		if(user.client.pixel_y >= 0) 
			view_y_offset = round(user.client.pixel_y/32)
		else 
			view_y_offset = Ceiling(user.client.pixel_y/32)
```

File: code/modules/clothing/glasses/night.dm [hunk 2]
```diff
@@ -124,76 +147,138 @@
...
+		for(var/target in blip_pool)
+			remove_marker(user, target, blip_pool[target])
+		STOP_PROCESSING(SSfastobj, src)
 
-	var/datum/action/item_action/m56_goggles/far_sight/FT = locate(/datum/action/item_action/m56_goggles/far_sight) in actions
-	if(FT)
-		FT.update_button_icon()
+	var/datum/action/item_action/m56_goggles/target_highlighting/TH = locate(/datum/action/item_action/m56_goggles/target_highlighting) in actions
+	if(TH)
+		TH.update_button_icon()
```
Comment (harryob) [inline code/modules/clothing/glasses/night.dm:177]:
```suggestion
	var/datum/action/item_action/m56_goggles/target_highlighting/action = locate(/datum/action/item_action/m56_goggles/target_highlighting) in actions
	if(action)
		action.update_button_icon()
```

File: code/modules/clothing/glasses/night.dm [hunk 3]
```diff
@@ -124,76 +147,138 @@
...
+		got_target = TRUE
+		show_marker(human_user, target)
+
+	if(got_target)
+		playsound_client(human_user.client, 'sound/items/tick.ogg')
+
+	for(var/target in blip_pool)
+		var/obj/effect/target_mark/TM = blip_pool[target]
+		if(target in viewed)
+			TM.update_marker_pos(human_user, target)
```
Comment (harryob) [inline code/modules/clothing/glasses/night.dm:236]:
```suggestion
		var/obj/effect/target_mark/mark = blip_pool[target]
		if(target in viewed)
			mark.update_marker_pos(human_user, target)
```

File: code/modules/clothing/glasses/night.dm [hunk 4]
```diff
@@ -124,76 +147,138 @@
...
+	if(!istype(human_user))
+		return
+
+	remove_marker(human_user, target, blip_pool[target])
+
+/obj/item/clothing/glasses/night/m56_goggles/proc/remove_marker(mob/user, atom/target, obj/effect/target_mark/TM)
+	if(user.client)
+		user.client.remove_from_screen(TM)
+		blip_pool -= target
+		qdel(TM)
```
Comment (harryob) [inline code/modules/clothing/glasses/night.dm:261]:
```suggestion
/obj/item/clothing/glasses/night/m56_goggles/proc/remove_marker(mob/user, atom/target, obj/effect/target_mark/mark)
	if(!user.client)
		return

	user.client.remove_from_screen(mark)
	blip_pool -= target
	qdel(mark)
```

File: code/modules/clothing/glasses/night.dm [hunk 5]
```diff
@@ -86,26 +86,49 @@
...
 	req_skill_level = null
 
+/obj/effect/target_mark
+	icon = 'icons/obj/items/marine-items.dmi'
+	icon_state = "tracker_blip"
+	layer = BELOW_FULLSCREEN_LAYER
+	plane = FULLSCREEN_PLANE
+
+/obj/effect/target_mark/proc/update_marker_pos(mob/user, atom/target)
+	var/c_view = user.client.view
```
Comment (harryob) [inline code/modules/clothing/glasses/night.dm:96]: can you guarantee there's always gonna be a client here? as byond clients have a tendency to be able to disable at pretty much any point during code execution

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/clothing/glasses/night.dm (hunk 5 above)
Comment (xDanilcusx) [reply code/modules/clothing/glasses/night.dm:96]: basically it's got a "!human_user.client" in start of the proc that calls it
only issue with that may be a 33 lines gap between check and actual proccall
dunno if this is alright or too unsafe
https://github.com/PvE-CMSS13/PvE-CMSS13/blob/30422dfb229cf5084275c1d47e9cf36c172cab00/code/modules/clothing/glasses/night.dm#L218

**Comment by xDanilcusx** [comment]
There is something bigger coming related to smartgun targeting system [including less lazy visuals], so this PR became pointless i guess.

**Pull request closed without merging**